pandas/core/series.py

```diff
@@ -3307,21 +3284,24 @@ def _try_kind_sort(arr):
         if not is_bool(ascending):
             raise ValueError("ascending must be boolean")
 
-        if not ascending:
-            argsorted = argsorted[::-1]
-
-        if na_position == "last":
-            n = good.sum()
-            sorted_index[:n] = idx[good][argsorted]
-            sorted_index[n:] = idx[bad]
-        elif na_position == "first":
-            n = bad.sum()
-            sorted_index[n:] = idx[good][argsorted]
-            sorted_index[:n] = idx[bad]
-        else:
-            raise ValueError(f"invalid na_position: {na_position}")
+        arr = self._values
 
-        result = self._constructor(arr[sorted_index], index=self.index[sorted_index])
+        if key:
+            if isinstance(arr, SparseArray):
+                # SparseArray doesn't store NaNs item-by-item, so pass everything
+                arr = ensure_key_mapped(self, key)._values
```

Comment:
SparseArrays can't be set by assignment, so just pass everything. Key func automatically runs on non-nans.

Reply:
hmm it would be nice to unify this

Reply:
I agree, but so far have been unable to. Would appreciate any suggestions.

The problem  is that SparseArray doesn't support setting elements, so you can't do `arr[good] = something`. The workarounds I was able to Google are far too ugly to use.

Reply:
why do you need to convert this to `_values` beforet he assignment? i think that would make this agnostic (and just select ._valuses when passing nargsort

Reply:
Thanks! Since we got rid of processing non-missing and missing values separately, this is now fixed.